pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -240,6 +248,62 @@ func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolc
 	return identity, false, nil
 }
 
+func (r *ReconcileUserAccount) ensureNSTemplateSet(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*toolchainv1alpha1.NSTemplateSet, bool, error) {
+	name := userAcc.Name
+
+	// create if not found
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name, Namespace: userAcc.Namespace}, nsTmplSet); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("creating a new NSTemplateSet", "name", name)
+			if err := r.setStatusProvisioning(userAcc); err != nil {
+				return nil, false, err
+			}
+			nsTmplSet = newNSTemplateSet(userAcc)
+			if err := controllerutil.SetControllerReference(userAcc, nsTmplSet, r.scheme); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to set controller reference for NSTemplateSet '%s'", name)
+			}
+			if err := r.client.Create(context.TODO(), nsTmplSet); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to create NSTemplateSet '%s'", name)
+			}
+			logger.Info("NSTemplateSet created successfully", "name", name)
+			return nsTmplSet, true, nil
+		}
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err, "failed to get NSTemplateSet '%s'", name)
+	}
+	logger.Info("NSTemplateSet already exists", "name", name)
+
+	// update if not same
+	equal := nsTmplSet.Spec.CompareTo(userAcc.Spec.NSTemplateSet)
+	if !equal {
+		logger.Info("updating NSTemplateSet", "name", name)
+		if err := r.setStatusProvisioning(userAcc); err != nil {
+			return nil, false, err
+		}
+		nsTmplSet.Spec = userAcc.Spec.NSTemplateSet
+		if err := r.client.Update(context.TODO(), nsTmplSet); err != nil {
+			return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err, "failed to update NSTemplateSet '%s'", name)
+		}
+		logger.Info("NSTemplateSet updated successfully", "name", name)
+		return nsTmplSet, true, nil
+	}
+
+	// update status if ready=false
+	readyCond, found := condition.FindConditionByType(nsTmplSet.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	if found {
+		if readyCond.Status == corev1.ConditionFalse {
+			if err := r.setStatusFromNSTemplateSet(userAcc, readyCond.Reason, readyCond.Message); err != nil {
+				return nsTmplSet, false, err
+			}
+			return nsTmplSet, true, nil
```

Comment:
there was nothing created nor updated except for status, so I don't think that we should return `true` here

anyway, I'm not sure if we really need to update the `UserAcc.status` every time when the `NsTemplateSet` is updated from one `status.ready=false` to another one where the `ready` is still `false`. I would say that we should do it the other way around - wait until the `status = true` and then update also the `UserAcc.status` to `true` as well

Reply:
> there was nothing created nor updated except for status, so I don't think that we should return true here

If we return false, this will get executed [link](https://github.com/codeready-toolchain/member-operator/blob/master/pkg/controller/useraccount/useraccount_controller.go#L147) and it will set UserAcc.Ready=true.

> I would say that we should do it the other way around - wait until the status = true and then update also the UserAcc.status to true as well

If we do other way then what if NSTmplSet.Ready remain `false` forever due to xyz_reason .. in this case UserAcc.Ready also stay `false` but xyz_reason will NOT be reflected in UserAcc status.

Reply:
> If we do other way then what if NSTmplSet.Ready remain false forever due to xyz_reason .. in this case UserAcc.Ready also stay false but xyz_reason will NOT be reflected in UserAcc status.

so if you set `UserAcc.Ready=true` only when `NsTemplateSet.status` is `true` as well, then it will fulfill your expectaction.

Reply:
please, keep in mind that every update of the `UserAcc.status` triggers synchronization of MUR and UserAcc which is something what we don't want to do for every change in `NsTemplateSet.status`. We are interested only in the final result

Reply:
@MatousJobanek You said ..
> so if you set UserAcc.Ready=true only when NsTemplateSet.status is true as well, then it will fulfill your expectaction.

@MatousJobanek you may not getting my below point ..
>If we do other way then what if NSTmplSet.Ready remain false forever due to xyz_reason .. in this case UserAcc.Ready also stay false but xyz_reason will NOT be reflected in UserAcc status.

Assume, `johnsmith-dev` namespace creation failed due to `ResourceQuotaUnavailable` (just assume some error condition due to which NSTmplSet stay Ready=false forever).  In this case, if we are only updating UserAcc.Status.Ready=True only when NSTmplSet.Status.Ready=True then in this case UserAcc.Status.Ready will remain "false" forever and reason would be "Provisioning".  So only changing UserAcc.Status.Ready=True will NOT be sufficient in this case.

Let me know in case I misunderstood or miss something.

Reply:
Let's first decide how we want user experience wrt UserAcc.Status and accordingly we will update UserAcc status.

Option1:
1. UserAcc.Status.Ready=false, Msg=Provisioning
2. NSTmplSet.Status.Ready=false, Msg=Provisioning dev
3. NSTmplSet.Status.Ready=false, Msg=Provisioning code
4. NSTmplSet.Status.Ready=true, Msg=Provisioned
5. UserAcc.Status.Ready=True, Msg=Provisioned

Option2:
1. UserAcc.Status.Ready=false, Msg=Provisioning
2. NSTmplSet.Status.Ready=false, Msg=Provisioning dev
3. UserAcc.Status.Ready=false, Msg=Provisioning dev
4. NSTmplSet.Status.Ready=false, Msg=Provisioning code
5. UserAcc.Status.Ready=false, Msg=Provisioning code
6. NSTmplSet.Status.Ready=true, Msg=Provisioned
7. UserAcc.Status.Ready=True, Msg=Provisioned

Option2 - error scenario will be reflected.  For ex, there is error in `dev` namespace provisioning then
- NSTmplSet.Status.Ready=false, Msg=Unable to provision `dev` namespace
- UserAcc.Status.Ready=false, Msg=Unable to provision `dev` namespace

While in case of Option1, above error scenario UserAcc.Status will NOT be updated so it will remain `UserAcc.Status.Ready=false, Msg=Provisioning`.

Also, there is Race condition possible.  Consider below events in timeline.

T1. UserAcc.Status.Ready=false, Msg=Provisioning
T2. NSTmplSet not exists, so create NSTmplSet
T3. UserAcc.Reconcile() return false
T4. UserAcc.Reconcile() called, as NSTmplSet exists everything is done and UserAcc.Status.Ready=True and Msg=Provisioned
T5. NSTmplSet.Reconcile() called (due to event at T2).  NSTmplSet.Status.Ready=false, Msg=Provisionng dev

No assume T5 happens just after T2.  So we need to decide status_update solution considering above race_condition.

In case of Opiton1, with above race_condition, UserAcc.Status.Ready will remain `True` even though namespace are getting provisioned.  But in case of Option2, for sometime UserAcc.Status.Ready will remain `True` and then later become `False` to reflect Namespaces are getting provisioned.

Reply:
I would go with the option 1. My reasoning:
* option 2 includes a lot of changes of `UserAcc.Status` that have a snowball effect and trigger synchronization between MUR <-> UserAcc.Status just for the sake of changing the condition msg. I think that for one provisioning flow it is too many operations
* if there is some issue, the `Reason` in both `UserAcc.Status` `NSTmplSet.Status` will still stay equal to `Provisioning` as the controller will still try to provision the namespace in a loop. There is no big difference between these two approaches - just the msg is different. For the first glance, it will be still in the `Provisioning` state.

But, if you still insist on propagating the error messages to MUR, then I have another option:

Option 3:
How about slightly modify the options and not update the message for every namespace separately and have one `Provisioning` message:
1. UserAcc.Status.Ready=false, Reason=Provisioning, Msg=Provisioning
2.  -> Provisioning dev
3.  -> Provisioning code
4. NSTmplSet.Status.Ready=true, Reason=Provisioned, Msg=Provisioned
5. UserAcc.Status.Ready=True, Reason=Provisioned, Msg=Provisioned

And for the error scenario:

1. UserAcc.Status.Ready=false, Reason=Provisioning, Msg=Provisioning
2.  -> Provisioning dev
3.  -> Provisioning code
4.  ----> error in provisioning code
5. UserAcc.Status.Ready=false, Reason=Provisioning, Msg=Unable to provision code namespace
6. UserAcc.Status.Ready=false, Reason=Provisioning, Msg=Unable to provision code namespace

Reply:
I don't think we can be too detailed in NSTmplSet.Status.Ready reason to specify what namespace is being provisioned and what status it has. We can get race condition when different reconcile loops triggered by different namespace changing the status reason. The same for the message.
So the reason has to be generic `Provisioning` and specify the message only if we got an error (that's fine if the error message will be cleaned by another reconcile loop. It will be re-added again if the error persists).
If we really want to track the status of individual namespace provisioning then we have to use dedicated conditions (one per NS type) and update them independently. It might be a good thing to do for better debugging but I would do it in a separate task/PR and keep it simple for now.

If we agree on the convention that we set the condition message for Ready only if there is an error then for UserAcc we could update the status only if the NSTemplateSet.Status.Ready.Message has been changed and ignore all other NSTemplateSet.Status updates if it's still Ready == false. Alternatively we would need to introduce an Error Condition in NSTemplateSet but it could really complicate the flow. So, I would rather stick with the simplest option (no UserAcc status updates until NSTemplateSet.Status.Ready == true) for now until we introduce dedicated conditions for each ns type in NSTemplateSet (see above) or check the NSTemplateSet messages to make updates in UserAcc.Status between Ready=fale,Reason=provisioning and Ready=true,Reason=provisioned in case of errors.

Reply:
If we are fine to agree upon convention/adhoc then we can do something like ..

- Set NSTmplSet.Status.Ready=false, Msg=namespace specific msg
- Check in UserAccCtrl, if (NSTmplSet.Status.Ready=false) && (Reason **!=** "Provisioning") then update UserAcc.Status from NSTmplSet.Status

In this way, NSTmplSet.Status.Msg will be able to show namespace specific message through which progress is monitored and there will be same no of status update for UserAcc (only in case of error or ready=true). Also, better maintainable, it's easy to set Msg when Ready=false for NSTmplSet.

Anyway, let's finalized on some way and we can do the changes accordingly.

Reply:
Yeah.. Looks good to me overall but I'm still not sure if we want to use messages for something else rather than for errors. And for errors we should use specific reasons. Not just "Provisioning":

1. When we are provisioning then we set NSTmplSet.Status.Ready=false, Reason=Provisioning, Message=""
2. When an error happened then we set NSTmplSet.Status.Ready=false, Reason=UnableToCreateNS/UnableToProcessNSTemplate/etc, Message=error_message
3. In UserAccount we can check the reason of NSTmplSet. If NSTmplSet.Status.Ready == false AND NSTmplSet.Status.Ready.Reason != Provisioning then update UserAcc.Status from NSTmplSet.Status. Or we can create a special Reason `type ErrorReason sting` then we can check the type instead of the value (Reason != Provisioning).

Reply:
Yeah, I agree with @alexeykazakov that we don't need to add the message for the `Provisioning` reason. Not sure why I added it there :-/ - sorry for that.
I believe that if we follow the rules that we:
* update the `NSTmplSet.Status` only when it's needed
* set message only for errors

then we can easily reflect all updates of `NSTmplSet.Status` in the `UserAcc.Status`. So I agree with the first two points in the last Alexey's comment. 
as for the third one. I'm not sure if such a check (or special error reason) is needed. When we introduce also the updates, then we would need to add an additional check for equality of the new `Updating` reason. After all, every time when there is an error the reason is changed too so we don't need to additionaly check that.
Anyway, for now, I believe that we can agree on option 3 from my comment - just without having messages for the Provisioning/Provisioned states and setting it only for errors, is that correct?

Reply:
Yeah right. The check I mentioned is probably redundant.
+1 to Matous’ last comment.

Reply:
Done https://github.com/codeready-toolchain/member-operator/pull/53/commits/1d29e1810e37bca659573ad33307d712807e4596